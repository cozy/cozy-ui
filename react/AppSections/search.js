import overEvery from 'lodash/overEvery'
import mapValues from 'lodash/mapValues'
import matches from 'lodash/matches'

import { APP_TYPE } from './constants'

const getDoctypesList = permsObj => {
  const doctypes = []
  for (let p in permsObj) {
    if (permsObj[p].type) doctypes.push(permsObj[p].type)
  }
  return doctypes
}

const typeMatcher = type => app => app.type === type
const categoryMatcher = category => app => {
  if (!Array.isArray(app.categories)) return false
  return app.categories.includes(category)
}
const tagMatcher = tag => app => {
  if (!Array.isArray(app.tags)) return false
  return app.tags.includes(tag)
}
const doctypeMatcher = doctype => app => {
  if (!app.permissions) return false
  return getDoctypesList(app.permissions).includes(doctype)
}
const pendingUpdateMatcher = () => app => !!app.availableVersion

const showMaintenanceMatcher = isShowMaintenance => app => {
  if (isShowMaintenance) return true
  return app.maintenance === undefined
}

const qualificationLabelsMatcher = label => app => {
  if (!Array.isArray(app.qualification_labels)) return false
  return app.qualification_labels.includes(label)
}

const searchAttrToMatcher = {
  type: typeMatcher,
  category: categoryMatcher,
  tag: tagMatcher,
  doctype: doctypeMatcher,
  pendingUpdate: pendingUpdateMatcher,
  showMaintenance: showMaintenanceMatcher,
  qualificationLabels: qualificationLabelsMatcher
}

/**
 * Returns a predicate function to match an app based on
 * a search specification.
 *
 * @param  {Search} search - What to search, ex: { type: 'webapp', category: 'banking'}
 * @return {Function<App>}
 */
export const makeMatcherFromSearch = (search = {}) => {
  // Create all predicates from the search object
  const predicates = Object.values(
    mapValues(search, (value, name) => {
      return searchAttrToMatcher[name]?.(value)
    })
  )
  // Return a function returning true if all predicates pass
  return overEvery(predicates)
}

/**
 * Returns a predicate to find the corresponding category option for
 * a search
 *
 * @param  {Search}
 * @return {Function<CategoryOption>}
 */
export const makeOptionMatcherFromSearch = (search = {}) => {
  const typeParam = search.type
  const categoryParam = search.category
  if (typeParam === APP_TYPE.KONNECTOR && !categoryParam) {
    return matches({ value: 'konnectors' })
  } else if (!typeParam && !categoryParam) {
    return matches({ value: 'all' })
  }
  return matches({ value: categoryParam, type: typeParam })
}

/**
 * Returns a Search from a category option
 *
 * @param  {CategoryOption}
 * @return {Search}
 */
export const makeSearchFromOption = option => {
  if (option.value === 'all') {
    return {}
  }
  const search = {}
  if (option.value === 'konnectors') {
    search.type = APP_TYPE.KONNECTOR
  } else {
    search.type = option.type
    search.category = option.value
  }
  return search
}
