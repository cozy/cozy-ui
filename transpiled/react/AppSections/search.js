import overEvery from 'lodash/overEvery';
import mapValues from 'lodash/mapValues';
import matches from 'lodash/matches';
import { APP_TYPE } from "cozy-ui/transpiled/react/AppSections/constants";

var getDoctypesList = function getDoctypesList(permsObj) {
  var doctypes = [];

  for (var p in permsObj) {
    if (permsObj[p].type) doctypes.push(permsObj[p].type);
  }

  return doctypes;
};

var typeMatcher = function typeMatcher(type) {
  return function (app) {
    return app.type === type;
  };
};

var categoryMatcher = function categoryMatcher(category) {
  return function (app) {
    if (!Array.isArray(app.categories)) return false;
    return app.categories.includes(category);
  };
};

var tagMatcher = function tagMatcher(tag) {
  return function (app) {
    if (!Array.isArray(app.tags)) return false;
    return app.tags.includes(tag);
  };
};

var doctypeMatcher = function doctypeMatcher(doctype) {
  return function (app) {
    if (!app.permissions) return false;
    return getDoctypesList(app.permissions).includes(doctype);
  };
};

var pendingUpdateMatcher = function pendingUpdateMatcher() {
  return function (app) {
    return !!app.availableVersion;
  };
};

var showMaintenanceMatcher = function showMaintenanceMatcher(isShowMaintenance) {
  return function (app) {
    if (isShowMaintenance) return true;
    return app.maintenance === undefined;
  };
};

var qualificationLabelsMatcher = function qualificationLabelsMatcher(label) {
  return function (app) {
    if (!Array.isArray(app.qualification_labels)) return false;
    return app.qualification_labels.includes(label);
  };
};

var searchAttrToMatcher = {
  type: typeMatcher,
  category: categoryMatcher,
  tag: tagMatcher,
  doctype: doctypeMatcher,
  pendingUpdate: pendingUpdateMatcher,
  showMaintenance: showMaintenanceMatcher,
  qualificationLabels: qualificationLabelsMatcher
};
/**
 * Returns a predicate function to match an app based on
 * a search specification.
 *
 * @param  {Search} search - What to search, ex: { type: 'webapp', category: 'banking'}
 * @return {Function<App>}
 */

export var makeMatcherFromSearch = function makeMatcherFromSearch() {
  var search = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  // Create all predicates from the search object
  var predicates = Object.values(mapValues(search, function (value, name) {
    var _searchAttrToMatcher$;

    return (_searchAttrToMatcher$ = searchAttrToMatcher[name]) === null || _searchAttrToMatcher$ === void 0 ? void 0 : _searchAttrToMatcher$.call(searchAttrToMatcher, value);
  })); // Return a function returning true if all predicates pass

  return overEvery(predicates);
};
/**
 * Returns a predicate to find the corresponding category option for
 * a search
 *
 * @param  {Search}
 * @return {Function<CategoryOption>}
 */

export var makeOptionMatcherFromSearch = function makeOptionMatcherFromSearch() {
  var search = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var typeParam = search.type;
  var categoryParam = search.category;

  if (typeParam === APP_TYPE.KONNECTOR && !categoryParam) {
    return matches({
      value: 'konnectors'
    });
  } else if (!typeParam && !categoryParam) {
    return matches({
      value: 'all'
    });
  }

  return matches({
    value: categoryParam,
    type: typeParam
  });
};
/**
 * Returns a Search from a category option
 *
 * @param  {CategoryOption}
 * @return {Search}
 */

export var makeSearchFromOption = function makeSearchFromOption(option) {
  if (option.value === 'all') {
    return {};
  }

  var search = {};

  if (option.value === 'konnectors') {
    search.type = APP_TYPE.KONNECTOR;
  } else {
    search.type = option.type;
    search.category = option.value;
  }

  return search;
};