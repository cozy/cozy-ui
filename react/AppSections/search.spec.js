'use strict'

/* eslint-env jest */

import { makeMatcherFromSearch } from './search'
import mockApps from '../mocks/apps'

const mockMaintenanceApps = [
  { slug: 'collect' },
  { slug: 'drive' },
  {
    slug: 'konnectorInMaintenance',
    maintenance: {
      flag_infra_maintenance: true,
      flag_short_maintenance: true,
      flag_disallow_manual_exec: true,
      messages: []
    }
  }
]

describe('makeMatcherFromSearch', () => {
  it('should filter correctly on type', () => {
    const matcher = makeMatcherFromSearch({ type: 'webapp' })
    expect(mockApps.filter(matcher)).toMatchSnapshot()
  })

  it('should filter correctly on doctype', () => {
    const matcher = makeMatcherFromSearch({ doctype: 'io.mock.doctype2' })
    expect(mockApps.filter(matcher)).toMatchSnapshot()
  })

  it('should filter correctly on tags', () => {
    const matcher = makeMatcherFromSearch({ tag: 'bills' })
    expect(mockApps.filter(matcher)).toMatchSnapshot()
  })

  it('should filter correctly on category', () => {
    const matcher = makeMatcherFromSearch({ category: 'cozy' })
    expect(mockApps.filter(matcher)).toMatchSnapshot()
  })

  it('should filter correctly on pending update apps', () => {
    const matcher = makeMatcherFromSearch({ pendingUpdate: true })
    expect(mockApps.filter(matcher)).toMatchSnapshot()
  })

  it('should filter correctly when show maintenance is false', () => {
    const matcher = makeMatcherFromSearch({ showMaintenance: false })
    expect(mockMaintenanceApps.filter(matcher)).toStrictEqual([
      { slug: 'collect' },
      { slug: 'drive' }
    ])
  })

  it('should filter correctly when show maintenance is true', () => {
    const matcher = makeMatcherFromSearch({ showMaintenance: true })
    expect(mockMaintenanceApps.filter(matcher)).toStrictEqual([
      { slug: 'collect' },
      { slug: 'drive' },
      {
        maintenance: {
          flag_disallow_manual_exec: true,
          flag_infra_maintenance: true,
          flag_short_maintenance: true,
          messages: []
        },
        slug: 'konnectorInMaintenance'
      }
    ])
  })

  it('should handle correctly multi filters', () => {
    const matcher = makeMatcherFromSearch({
      type: 'konnector',
      doctype: 'io.mock.doctype'
    })
    expect(mockApps.filter(matcher)).toMatchSnapshot()
  })

  it('should not throw error if search is not provided', () => {
    const matcher = makeMatcherFromSearch()
    expect(mockApps.filter(matcher)).toMatchSnapshot()
  })

  it('should not throw error if app with malformed permission and search on doctype', () => {
    const malFormedPermApp = {
      slug: 'collect',
      name: 'Collect',
      editor: 'Cozy',
      name_prefix: 'Cozy',
      categories: ['cozy'],
      developer: { name: 'Cozy' },
      type: 'webapp',
      icon: '<svg></svg>',
      permissions: {
        mock: {
          // NO TYPE = malformed
          description: 'io.mock.doctype'
        },
        mock2: {
          type: 'io.mock.doctype2'
        }
      },
      tags: ['konnector', 'collect', 'bills', 'providers', 'files'],
      version: '3.0.0',
      versions: {
        stable: ['3.0.0'],
        beta: ['3.0.0'],
        dev: ['3.0.0']
      },
      uninstallable: false,
      installed: true,
      isInRegistry: true,
      related: 'http://collect.cozy.mock/'
    }
    const matcher = makeMatcherFromSearch({ doctype: 'io.mock.doctype' })
    expect([malFormedPermApp].filter(matcher)).toMatchSnapshot()
  })

  it('should not throw error if app with no categories and search on category', () => {
    const appWithoutCategories = {
      slug: 'collect',
      name: 'Collect',
      editor: 'Cozy',
      name_prefix: 'Cozy',
      developer: { name: 'Cozy' },
      type: 'webapp',
      icon: '<svg></svg>',
      permissions: {
        mock: {
          type: 'io.mock.doctype'
        },
        mock2: {
          type: 'io.mock.doctype2'
        }
      },
      tags: ['konnector', 'collect', 'bills', 'providers', 'files'],
      version: '3.0.0',
      versions: {
        stable: ['3.0.0'],
        beta: ['3.0.0'],
        dev: ['3.0.0']
      },
      uninstallable: false,
      installed: true,
      isInRegistry: true,
      related: 'http://collect.cozy.mock/'
    }
    const matcher = makeMatcherFromSearch({ category: 'cozy' })
    expect([appWithoutCategories].filter(matcher)).toMatchSnapshot()
  })

  it('should work even for an unsupported filter', () => {
    const matcher = makeMatcherFromSearch({ unsupportedKey: 'omg' })
    expect(mockApps.filter(matcher)).toMatchSnapshot()
  })
})
