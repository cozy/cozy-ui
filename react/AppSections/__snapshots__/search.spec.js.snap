// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeMatcherFromSearch should filter correctly on category 1`] = `
Array [
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "label": 1,
    "name": "Collect",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://collect.cozy.mock/",
    "slug": "collect",
    "tags": Array [
      "konnector",
      "collect",
      "bills",
      "providers",
      "files",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "name": "Drive",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://drive.cozy.mock/",
    "slug": "drive",
    "tags": Array [
      "search",
      "files",
      "folders",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0-beta89bnhj3993",
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "locales": Object {
      "en": Object {
        "long_description": "A long description finally short",
      },
    },
    "name": "Photos",
    "name_prefix": "Cozy",
    "slug": "photos",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
]
`;

exports[`makeMatcherFromSearch should filter correctly on doctype 1`] = `
Array [
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "label": 1,
    "name": "Collect",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://collect.cozy.mock/",
    "slug": "collect",
    "tags": Array [
      "konnector",
      "collect",
      "bills",
      "providers",
      "files",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "name": "Drive",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://drive.cozy.mock/",
    "slug": "drive",
    "tags": Array [
      "search",
      "files",
      "folders",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0-beta89bnhj3993",
  },
]
`;

exports[`makeMatcherFromSearch should filter correctly on pending update apps 1`] = `
Array [
  Object {
    "availableVersion": "3.0.0",
    "categories": Array [
      "partners",
    ],
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Tasky",
    "slug": "tasky",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "3.0.0",
        "1.0.0",
      ],
      "dev": Array [
        "3.0.0",
        "1.0.0",
      ],
      "stable": Array [
        "3.0.0",
        "1.0.0",
      ],
    },
  },
]
`;

exports[`makeMatcherFromSearch should filter correctly on tags 1`] = `
Array [
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "label": 1,
    "name": "Collect",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://collect.cozy.mock/",
    "slug": "collect",
    "tags": Array [
      "konnector",
      "collect",
      "bills",
      "providers",
      "files",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "transport",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Trinlane",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
    },
    "slug": "konnector-trinlane",
    "tags": Array [
      "transport",
      "files",
      "bills",
    ],
    "type": "konnector",
    "uninstallable": true,
    "version": "0.1.0",
    "versions": Object {
      "beta": Array [
        "0.1.0",
      ],
      "dev": Array [
        "0.1.0",
      ],
      "stable": Array [
        "0.1.0",
      ],
    },
  },
]
`;

exports[`makeMatcherFromSearch should filter correctly on type 1`] = `
Array [
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "label": 1,
    "name": "Collect",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://collect.cozy.mock/",
    "slug": "collect",
    "tags": Array [
      "konnector",
      "collect",
      "bills",
      "providers",
      "files",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "others",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "locales": Object {
      "name": "For Dev Only",
    },
    "name": "DevOnly",
    "slug": "devonly",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [],
      "dev": Array [
        "1.0.0-betaojdkehy989ekhflldh",
      ],
      "stable": Array [],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "name": "Drive",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://drive.cozy.mock/",
    "slug": "drive",
    "tags": Array [
      "search",
      "files",
      "folders",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0-beta89bnhj3993",
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "locales": Object {
      "en": Object {
        "long_description": "A long description finally short",
      },
    },
    "name": "Photos",
    "name_prefix": "Cozy",
    "slug": "photos",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "availableVersion": "3.0.0",
    "categories": Array [
      "partners",
    ],
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Tasky",
    "slug": "tasky",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "3.0.0",
        "1.0.0",
      ],
      "dev": Array [
        "3.0.0",
        "1.0.0",
      ],
      "stable": Array [
        "3.0.0",
        "1.0.0",
      ],
    },
  },
]
`;

exports[`makeMatcherFromSearch should handle correctly multi filters 1`] = `
Array [
  Object {
    "categories": Array [
      "transport",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Trinlane",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
    },
    "slug": "konnector-trinlane",
    "tags": Array [
      "transport",
      "files",
      "bills",
    ],
    "type": "konnector",
    "uninstallable": true,
    "version": "0.1.0",
    "versions": Object {
      "beta": Array [
        "0.1.0",
      ],
      "dev": Array [
        "0.1.0",
      ],
      "stable": Array [
        "0.1.0",
      ],
    },
  },
]
`;

exports[`makeMatcherFromSearch should not throw error if app with malformed permission and search on doctype 1`] = `Array []`;

exports[`makeMatcherFromSearch should not throw error if app with no categories and search on category 1`] = `Array []`;

exports[`makeMatcherFromSearch should not throw error if search is not provided 1`] = `
Array [
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "label": 1,
    "name": "Collect",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://collect.cozy.mock/",
    "slug": "collect",
    "tags": Array [
      "konnector",
      "collect",
      "bills",
      "providers",
      "files",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "others",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "locales": Object {
      "name": "For Dev Only",
    },
    "name": "DevOnly",
    "slug": "devonly",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [],
      "dev": Array [
        "1.0.0-betaojdkehy989ekhflldh",
      ],
      "stable": Array [],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "name": "Drive",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://drive.cozy.mock/",
    "slug": "drive",
    "tags": Array [
      "search",
      "files",
      "folders",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0-beta89bnhj3993",
  },
  Object {
    "categories": Array [
      "isp",
      "telecom",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "name": "Bouilligue",
    "slug": "konnector-bouilligue",
    "type": "konnector",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "0.1.0",
      ],
      "dev": Array [
        "0.1.0",
      ],
      "stable": Array [
        "0.1.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "transport",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Trinlane",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
    },
    "slug": "konnector-trinlane",
    "tags": Array [
      "transport",
      "files",
      "bills",
    ],
    "type": "konnector",
    "uninstallable": true,
    "version": "0.1.0",
    "versions": Object {
      "beta": Array [
        "0.1.0",
      ],
      "dev": Array [
        "0.1.0",
      ],
      "stable": Array [
        "0.1.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "locales": Object {
      "en": Object {
        "long_description": "A long description finally short",
      },
    },
    "name": "Photos",
    "name_prefix": "Cozy",
    "slug": "photos",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "availableVersion": "3.0.0",
    "categories": Array [
      "partners",
    ],
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Tasky",
    "slug": "tasky",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "3.0.0",
        "1.0.0",
      ],
      "dev": Array [
        "3.0.0",
        "1.0.0",
      ],
      "stable": Array [
        "3.0.0",
        "1.0.0",
      ],
    },
  },
]
`;

exports[`makeMatcherFromSearch should work even for an unsupported filter 1`] = `
Array [
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "label": 1,
    "name": "Collect",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://collect.cozy.mock/",
    "slug": "collect",
    "tags": Array [
      "konnector",
      "collect",
      "bills",
      "providers",
      "files",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "others",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "locales": Object {
      "name": "For Dev Only",
    },
    "name": "DevOnly",
    "slug": "devonly",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [],
      "dev": Array [
        "1.0.0-betaojdkehy989ekhflldh",
      ],
      "stable": Array [],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "name": "Drive",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://drive.cozy.mock/",
    "slug": "drive",
    "tags": Array [
      "search",
      "files",
      "folders",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0-beta89bnhj3993",
  },
  Object {
    "categories": Array [
      "isp",
      "telecom",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "name": "Bouilligue",
    "slug": "konnector-bouilligue",
    "type": "konnector",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "0.1.0",
      ],
      "dev": Array [
        "0.1.0",
      ],
      "stable": Array [
        "0.1.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "transport",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Trinlane",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
    },
    "slug": "konnector-trinlane",
    "tags": Array [
      "transport",
      "files",
      "bills",
    ],
    "type": "konnector",
    "uninstallable": true,
    "version": "0.1.0",
    "versions": Object {
      "beta": Array [
        "0.1.0",
      ],
      "dev": Array [
        "0.1.0",
      ],
      "stable": Array [
        "0.1.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "locales": Object {
      "en": Object {
        "long_description": "A long description finally short",
      },
    },
    "name": "Photos",
    "name_prefix": "Cozy",
    "slug": "photos",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "availableVersion": "3.0.0",
    "categories": Array [
      "partners",
    ],
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Tasky",
    "slug": "tasky",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "3.0.0",
        "1.0.0",
      ],
      "dev": Array [
        "3.0.0",
        "1.0.0",
      ],
      "stable": Array [
        "3.0.0",
        "1.0.0",
      ],
    },
  },
]
`;
